proposals/params-collections.md

```diff
@@ -341,6 +341,45 @@ If only a single candidate passes the test, the invocation of the candidate is s
 - the candidate is either not generic, or its type arguments are explicitly specified;
 - there is no ambiguity between normal and expanded forms of the candidate that cannot be resolved at compile time. 
 
+In this case, result of the invocation matches result of a regular invocation of the single candidate,
+except in a situation when all the following conditions are met. The result is the returned value converted to `dynamic` then.
+- the candidate is not a local function;
+- the candidate returns a value (doesn't have type `void`, doesn't return a `ref`);
+- there is an implicit conversion from result type to `dynamic`;
+- the receiver and argument list would be supported for dynamically bound invocation.
+
+For example:
+```csharp
+unsafe public class C
+{
+    public static void Main()
+    {
+        dynamic d = 1;
+        var a = Test1(d); // error CS0815: Cannot assign void to an implicitly-typed variable
+        var b = Test2(d); // void*, there is no implicit conversion to dynamic
+        var c = Test3(d); // dynamic
+        c.InvokeOnDynamic(); // No error
+
+        var e = test4(d); // int
+        static int test4(int x) => x;
+
+        var f = Test5(d); // int
```

Comment:
Shouldn't this be an error given it's a `ref` return?

Reply:
To clarify, consider the following example:
```
public class C
{
    public static void Main()
    {
        dynamic d = 1;
        var a = Test1(d);
        System.Console.WriteLine(a);        
    }

    static int _test1 = 0;    
    static ref int Test1(int x) => ref _test1;
}
```

When compiled in C# 12, 'a' is dynamic and the code produces the following output:
```
System.InvalidCastException: The result type 'System.Int32&' of the dynamic binding produced by binder 'Microsoft.CSharp.RuntimeBinder.CSharpInvokeMemberBinder' is not compatible with the result type 'System.Object' expected by the call site.
   at System.Dynamic.DynamicMetaObjectBinder.Bind(Object[] args, ReadOnlyCollection`1 parameters, LabelTarget returnLabel)
   at System.Runtime.CompilerServices.CallSiteBinder.BindCore[T](CallSite`1 site, Object[] args)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
   at C.Main()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

When compiled in preview (prior to this design change), 'a' is int and the code produces the following output:
```
0
```

With the proposed change in design, there will be no change in behavior by comparison to the preview behavior.